contrib/kafka/pkg/dispatcher/dispatcher.go

```diff
@@ -131,12 +135,35 @@ func (d *KafkaDispatcher) UpdateConfig(config *multichannelfanout.Config) error
 			}
 		}
 
+		hcMap, err := createHostToChannelMap(config)
```

Comment:
`hostToChannelMap` and `kafkaConsumers` could go out of sync either ways. This is because `subscribe`
 and `unsubscribe` could both error out. Existing code has an issue that it doesn't handle the error. It should handle the error and return back so that reconcilliation will call it again. In current implementation if `subscribe` and `unsubscribe` fails due to intermittent issues such as couldn't connect to kafka endpoint due to network glitch, then it will never recover till some channel gets updated.

I think the question is more about, if `createHostToChannelMap` fails then should we or should we not update `kafkaConsumers`. The reason I chose to call 'createHostToChannelMap' at the end is because I know that it can fail only it there is a duplicate 'hostName', which can happen due to a bug in controller code (and less likely as there is a unit test for it). There are no external dependencies and hence likelyhood of it failing is low. Whereas the likelyhood of `subscribe/unsubscribe` failing is high due to external dependencies. BTW, this reasoning holds good only once we fix the code and handle the errors from `subscribe/unsubscribe`.
Let me know what you think.

Reply:
I agree with your reasoning, but it only applies to `setHostToChannelMap`, not `createHostToChannelMap`. As `createHostToChannelMap` can return an error (but doesn't mutate any state), run it before mutating any state. Then call `subscribe` and `unsubscribe` followed by `setHostToChannelMap`.

Although this raises a fairly important point -- we are ignoring the errors from subscribe and unsubscribe. We shouldn't. If we need to keep hostToChannelMap in-sync with the real subscriptions, should we update the map one entry at a time, just after subscribe/unsubscribe?

Reply:
Based on our offline conversation.
1. Moved the createHostToChannelMap before updating kafkaConsumers
2. kept setHostToChannelMap at the end of the loop
3. Created a bug #1072